alert-common/src/main/java/com/synopsys/integration/alert/common/workflow/processor/DigestMessageContentProcessor.java

```diff
@@ -23,91 +23,119 @@
+                                                        .flatMap(SortedSet::stream)
+                                                        .collect(Collectors.toList());
+        return combineCategoryItems(allCategoryItems);
+    }
+
+    private SortedSet<CategoryItem> combineCategoryItems(final List<CategoryItem> allCategoryItems) {
+        // The amount of collapsing we do makes this impossible to map back to a single notification.
+        final Long unknownNotificationId = Long.MIN_VALUE;
+        final Map<CategoryKey, CategoryItem> keyToItems = new LinkedHashMap<>();
+        for (final CategoryItem categoryItem : allCategoryItems) {
+            final CategoryKey categoryKey = generateCategoryKey(categoryItem);
+            final CategoryItem oldItem = keyToItems.get(categoryKey);
+
+            final SortedSet<LinkableItem> linkableItems;
+            if (null != oldItem) {
+                linkableItems = combineLinkableItems(oldItem.getItems(), categoryItem.getItems());
             } else {
-                categoryDataCache.put(key, categoryItem);
+                linkableItems = categoryItem.getItems();
             }
-            return null;
-        };
-    }
 
-    private void processOperation(final Map<CategoryKey, CategoryItem> categoryDataCache, final CategoryItem item) {
-        final ItemOperation operation = item.getOperation();
-        if (operationFunctionMap.containsKey(operation)) {
-            final BiFunction<Map<CategoryKey, CategoryItem>, CategoryItem, Void> operationFunction = operationFunctionMap.get(operation);
-            operationFunction.apply(categoryDataCache, item);
+            final CategoryItem newCategoryItem = new CategoryItem(categoryKey, categoryItem.getOperation(), unknownNotificationId, linkableItems);
+            keyToItems.put(categoryKey, newCategoryItem);
         }
+        return new TreeSet<>(keyToItems.values());
     }
 
-    private Optional<AggregateMessageContent> rebuildTopic(final AggregateMessageContent currentContent, final Collection<CategoryItem> categoryItemCollection) {
-        if (categoryItemCollection.isEmpty()) {
-            return Optional.empty();
-        } else {
-            final String url = currentContent.getUrl().orElse(null);
-            final LinkableItem subTopic = currentContent.getSubTopic().orElse(null);
-            return Optional.of(new AggregateMessageContent(currentContent.getName(), currentContent.getValue(), url, subTopic, new TreeSet<>(categoryItemCollection)));
+    private CategoryKey generateCategoryKey(final CategoryItem categoryItem) {
+        final List<String> keyParts = new ArrayList<>();
+        keyParts.add(categoryItem.getOperation().name());
+        for (final LinkableItem item : categoryItem.getItems()) {
+            if (!item.isCollapsible()) {
+                keyParts.add(item.getName());
+                keyParts.add(item.getValue());
+            }
         }
+
+        return CategoryKey.from("digest", keyParts);
```

Comment:
Do we have or want a constant for digest?

Reply:
I'm not sure. This string could be anything, my choice to use "digest" was mostly just because that was the class this was in. It will only be used internally. Does that still make it worth it?